python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py

```diff
@@ -1622,6 +1623,69 @@ def get_materialization_count_by_partition(
 
         return materialization_count_by_partition
 
+    def _check_partitions_table(self):
+        # Guards against cases where the user is not running the latest migration for
+        # partitions storage. Should be updated when the partitions storage schema changes.
+        if not self.has_table("mutable_partitions"):
+            raise DagsterInvalidInvocationError(
+                "Cannot add partitions to non-existent table. Add this table by running `dagster"
+                " instance migrate`."
+            )
+
+    def _fetch_partition_keys_for_partition_def(self, partitions_def_name: str) -> Sequence[str]:
+        columns = [
+            MutablePartitionsTable.c.partitions_def_name,
+            MutablePartitionsTable.c.partition_key,
+        ]
+        query = db.select(columns).where(
+            MutablePartitionsTable.c.partitions_def_name == partitions_def_name
+        )
+        with self.index_connection() as conn:
+            rows = conn.execute(query).fetchall()
+
+        return [row[1] for row in rows]
+
+    def get_partitions(self, partitions_def_name: str) -> Sequence[str]:
+        """Get the list of partition keys for a partition definition."""
+        self._check_partitions_table()
+        return self._fetch_partition_keys_for_partition_def(partitions_def_name)
+
+    def has_partition(self, partitions_def_name: str, partition_key: str) -> bool:
+        self._check_partitions_table()
+        return partition_key in self._fetch_partition_keys_for_partition_def(partitions_def_name)
+
+    def add_partitions(self, partitions_def_name: str, partition_keys: Sequence[str]) -> None:
+        if isinstance(partition_keys, str):
+            # Guard against a single string being passed in `partition_keys`
+            raise DagsterInvalidInvocationError("partition_keys must be a sequence of strings")
+        check.sequence_param(partition_keys, "partition_keys")
+
+        self._check_partitions_table()
+        existing_partitions = set(self.get_partitions(partitions_def_name))
```

Comment:
I think we should find a way to avoid fetching the entire set of partitions when adding a partition.